crates/iceberg/src/spec/table_metadata_builder.rs

```diff
@@ -0,0 +1,2070 @@
+    pub fn remove_properties(mut self, properties: &[String]) -> Self {
+        for property in properties {
+            self.metadata.properties.remove(property);
+        }
+
+        if !properties.is_empty() {
+            self.changes.push(TableUpdate::RemoveProperties {
+                removals: properties.to_vec(),
+            });
+        }
+
+        self
+    }
+
+    /// Set the location of the table metadata, stripping any trailing slashes.
+    pub fn set_location(mut self, location: String) -> Self {
+        let location = location.trim_end_matches('/').to_string();
+        if self.metadata.location != location {
+            self.changes.push(TableUpdate::SetLocation {
+                location: location.clone(),
+            });
+            self.metadata.location = location;
+        }
+
+        self
+    }
+
+    /// Add a snapshot to the table metadata.
+    ///
+    /// # Errors
+    /// - No schema has been added to the table metadata.
+    /// - No partition spec has been added to the table metadata.
+    /// - No sort order has been added to the table metadata.
+    /// - Snapshot id already exists.
+    /// - For format version > 1: the sequence number of the snapshot is lower than the highest sequence number specified so far.
+    pub fn add_snapshot(mut self, snapshot: Snapshot) -> Result<Self> {
+        if self
+            .metadata
+            .snapshots
+            .contains_key(&snapshot.snapshot_id())
+        {
+            return Err(Error::new(
+                ErrorKind::DataInvalid,
+                format!("Snapshot already exists for: '{}'", snapshot.snapshot_id()),
+            ));
+        }
+
+        if self.metadata.format_version != FormatVersion::V1
+            && snapshot.sequence_number() <= self.metadata.last_sequence_number
+            && snapshot.parent_snapshot_id().is_some()
+        {
+            return Err(Error::new(
+                ErrorKind::DataInvalid,
+                format!(
+                    "Cannot add snapshot with sequence number {} older than last sequence number {}",
+                    snapshot.sequence_number(),
+                    self.metadata.last_sequence_number
+                )
+            ));
+        }
+
+        if let Some(last) = self.metadata.snapshot_log.last() {
+            // commits can happen concurrently from different machines.
+            // A tolerance helps us avoid failure for small clock skew
+            if snapshot.timestamp_ms() - last.timestamp_ms < -ONE_MINUTE_MS {
+                return Err(Error::new(
+                    ErrorKind::DataInvalid,
+                    format!(
+                        "Invalid snapshot timestamp {}: before last snapshot timestamp {}",
+                        snapshot.timestamp_ms(),
+                        last.timestamp_ms
+                    ),
+                ));
+            }
+        }
+
+        if snapshot.timestamp_ms() - self.metadata.last_updated_ms < -ONE_MINUTE_MS {
```

Comment:
Just adding that we also check this when deserializing Metadata:
https://github.com/apache/iceberg-rust/blob/50345196c87b00badc1a6490aef284e84f4c3e9a/crates/iceberg/src/spec/table_metadata.rs#L370-L371

We also validate chronological timestamps of metadata logs.
I added those checks both here and in the `try_normalize` function because its in Java.

Should we decide to remove it here, we should also remove the checks that I linked above.

Generally I think a rough clock-skew check reduces confusement of users if one clock is way off. I don't think there is a strickt technical requirement for this?